minidump-processor/src/processor.rs

```diff
@@ -568,66 +568,81 @@ where
-                            });
-
-                        crate::ExceptionInfo {
-                            reason,
-                            address,
-                            adjusted_address,
-                            instruction_str: Some(op_analysis.instruction_str),
-                            memory_accesses: op_analysis.memory_accesses,
-                            possible_bit_flips: Default::default(),
-                        }
-                    })
-                    .map_err(|e| tracing::warn!("failed to analyze the thread context: {}", e))
-                    .ok()
-            })
-            .unwrap_or(crate::ExceptionInfo {
+        exception_context = exception.context(&dump_system_info, misc_info.as_ref());
+
+        // If we have a context, we can attempt to analyze the crashing thread's instructions
+        if let Some(context) = &exception_context {
+            match crate::op_analysis::analyze_thread_context(
+                context,
+                &memory_list,
+                stack_memory_ref,
+            ) {
+                Ok(op_analysis) => {
+                    let memory_accesses = op_analysis.memory_accesses.as_deref();
+
+                    let adjusted_address = try_detect_null_pointer_in_disguise(memory_accesses)
+                        .map(|offset| AdjustedAddress::NullPointerWithOffset(offset.into()))
+                        .or_else(|| {
+                            try_get_non_canonical_crash_address(
+                                &system_info,
+                                memory_accesses,
+                                reason,
+                                address,
+                            )
+                            .map(|addr| AdjustedAddress::NonCanonical(addr.into()))
+                        });
+
+                    exception_info = Some(crate::ExceptionInfo {
+                        reason,
+                        address: address.into(),
+                        adjusted_address,
+                        instruction_str: Some(op_analysis.instruction_str),
+                        memory_accesses: op_analysis.memory_accesses,
+                        possible_bit_flips: Default::default(),
+                    });
+                    instruction_registers = op_analysis.registers;
+                }
+                Err(e) => {
+                    tracing::warn!("failed to analyze the thread context: {e}");
+                }
+            }
+        }
+
+        if exception_info.is_none() {
+            exception_info = Some(crate::ExceptionInfo {
                 reason,
-                address,
+                address: address.into(),
                 adjusted_address: None,
                 instruction_str: None,
                 memory_accesses: None,
                 possible_bit_flips: Default::default(),
-            })
-    });
+            });
+        }
+    }
 
     // Check for bit-flips of the exception address
```

Comment:
Likewise, we could drop the comment and make a tersely-named function instead.

Reply:
I will do so, but there's so much contextual stuff to pass around that this function may be better refactored into a struct with methods in the future (though I won't do that now).

Reply:
I ended up refactoring it into a struct :shrug: